src/leveled_eval.erl

```diff
@@ -0,0 +1,648 @@
+    ?assertMatch(<<"SMITH">>, maps:get(<<"fn">>, EvalOut6)),
+    ?assertMatch(<<"19861216">>, maps:get(<<"dob">>, EvalOut6)),
+    ?assertMatch(<<"">>, maps:get(<<"dod">>, EvalOut6)),
+    ?assertMatch(<<"LS1 4BT#LS8 1ZZ">>, maps:get(<<"pcs">>, EvalOut6)),
+    ?assertMatch([<<"MA">>, <<"N1">>, <<"Ve">>], maps:get(<<"gns">>, EvalOut6)),
+
+    EvalOut7 =
+        apply_eval(
+            ParsedExp6,
+            <<"SMITH|19861216||MAN1VeZ|LS1 4BT#LS8 1ZZ">>,
+            <<"9000000001">>,
+            maps:new()
+        ),
+    ?assertMatch(<<"SMITH">>, maps:get(<<"fn">>, EvalOut7)),
+    ?assertMatch(<<"19861216">>, maps:get(<<"dob">>, EvalOut7)),
+    ?assertMatch(<<"">>, maps:get(<<"dod">>, EvalOut7)),
+    ?assertMatch(<<"LS1 4BT#LS8 1ZZ">>, maps:get(<<"pcs">>, EvalOut7)),
+    ?assertMatch([<<"MA">>, <<"N1">>, <<"Ve">>], maps:get(<<"gns">>, EvalOut7)),
+
+    EvalString8 = EvalString1 ++ " | split($gns, \"#\", $gns)",
+    {ok, Tokens8, _EndLine8} = leveled_evallexer:string(EvalString8),
+    {ok, ParsedExp8} = leveled_evalparser:parse(Tokens8),
+    EvalOut8 =
+        apply_eval(
+            ParsedExp8,
+            <<"SMITH|19861216||Willow#Mia#Vera|LS1 4BT#LS8 1ZZ">>,
+            <<"9000000001">>,
+            maps:new()
+        ),
+    ?assertMatch(<<"SMITH">>, maps:get(<<"fn">>, EvalOut8)),
+    ?assertMatch(<<"19861216">>, maps:get(<<"dob">>, EvalOut8)),
+    ?assertMatch(<<"">>, maps:get(<<"dod">>, EvalOut8)),
+    ?assertMatch(<<"LS1 4BT#LS8 1ZZ">>, maps:get(<<"pcs">>, EvalOut8)),
+    ?assertMatch([<<"Willow">>, <<"Mia">>, <<"Vera">>], maps:get(<<"gns">>, EvalOut8)),
+
+    EvalString9 =
+        "delim($term, \"|\", ($name, $height, $weight, $pick)) |"
+        " to_integer($height, $height) |"
+        " to_integer($weight, $weight) |"
+        " to_integer($pick, $pick) |"
+        " delim($key, \"|\", ($team, $number)) |"
+        " index($team, 0, 9, $doh)",
+    {ok, Tokens9, _EndLine9} = leveled_evallexer:string(EvalString9),
+    {ok, ParsedExp9} = leveled_evalparser:parse(Tokens9),
+    EvalOut9 =
+        apply_eval(
+            ParsedExp9,
+            <<"WEMBANYAMA|224cm|95kg|#1">>,
+            <<"SPURS|00001">>,
+            maps:new()
+        ),
+    ?assertMatch(<<"WEMBANYAMA">>, maps:get(<<"name">>, EvalOut9)),
+    ?assertMatch(224, maps:get(<<"height">>, EvalOut9)),
+    ?assertMatch(95, maps:get(<<"weight">>, EvalOut9)),
+    ?assertMatch(<<"#1">>, maps:get(<<"pick">>, EvalOut9)),
+        % Not changes as not starting with integer
+    ?assertMatch(<<"SPURS">>, maps:get(<<"team">>, EvalOut9)),
+    ?assertMatch(<<"00001">>, maps:get(<<"number">>, EvalOut9)),
+    ?assertNot(maps:is_key(<<"doh">>, EvalOut9)),
+
+    %% Age at 30 April 2024
+    EvalString10 =
+        EvalString5 ++ 
+        " | index($dob, 4, 4, $birthday)"
+        " | map($birthday, <=, ((\"0430\", 2024)), 2023, $yoc)"
+        " | subtract($yoc, $yob, $age)"
+        " | add($age, 1, $age_next)"
+        " | to_string($age, $age)"
```

Comment:
Didn't realize one can overload variables to have different types... probably not the nicest thing to use in an example.